driver/src/orderbook/streamed/orderbook.rs

```diff
@@ -0,0 +1,84 @@
+use super::*;
+use crate::{
+    contracts::stablex_contract::batch_exchange,
+    models::{AccountState, Order},
+    orderbook::StableXOrderBookReading,
+};
+use anyhow::Result;
+use ethcontract::{contract::EventData, H256, U256};
+use state::{Batch, State};
+use std::collections::BTreeMap;
+
+// Ethereum events (logs) can be both created and removed. Removals happen if the chain reorganizes
+// and ends up not including block that was previously thought to be part of the chain.
+// However, the orderbook state (`State`) cannot remove events. To support this, we keep an ordered
+// list of all events based on which the state is built.
+
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct EventSortKey {
+    block_number: u64,
+    /// Is included to differentiate events from the same block number but different blocks which
+    /// can happen during reorgs.
+    block_hash: H256,
+    log_index: usize,
+}
+
+#[derive(Debug)]
+struct Value {
+    event: batch_exchange::Event,
+    /// The batch id is calculated based on the timestamp of the block.
+    batch_id: BatchId,
+}
+
+pub struct Orderbook {
+    events: BTreeMap<EventSortKey, Value>,
+}
+
+impl Orderbook {
+    pub fn new() -> Self {
+        // TODO: Set up a thread that applies past events and listens for new events.
+        unimplemented!();
+    }
+
+    fn handle_event_data(
+        &mut self,
+        event_data: EventData<batch_exchange::Event>,
+        block_number: u64,
+        log_index: usize,
+        block_hash: H256,
+        block_timestamp: u64,
```

Comment:
This might be problematic as querying the node for all past time stamps can be really slow.

Reply:
I don't see a way around it. I have tested this on rinkeby by querying the node for all timestamps needed and caching the last timestamp to speed up multiple events from the same block.

Reply:
The only way I see would be to handle past events slightly differently than future events, so for past events you apply all the events under the current batch ID (queried from the smart contract for example) and then for incoming events you can query the block timestamp for each (since they will be way more sparse).

Definitely a separate issue/PR though.